### PR TITLE
Unified BufferedCSVReaderOptions parsing

### PR DESCRIPTION
**Pull request by Tishj**
This PR removes* the code focused around parsing `options` that is scattered around `copy_csv` and `read_csv` and replaces them with new methods defined in the `BufferedCSVReaderOptions`.

The options have been divided into three parts, Common, Read and Write.
- Read - parsed by `SetReadOption`.
- Write -  parsed by `SetWriteOption`.
- Common - parsed by `SetBaseOption` which is called by both `SetReadOption` and `SetWriteOption`.

I've added an assertion to `SetBaseOption` to ensure it's only given a lowercase-converted version of the option string.

\* the "columns" option is still parsed inside of ReadCSVBind, as it deals with `return_types` and I felt it was too different from the other cases so it should remain there.  
If this exception is misplaced, I'll also unify this case into the `SetReadOption` method.

### CODE REVIEW

File: src/function/table/read_csv.cpp [hunk 1]
```diff
@@ -30,78 +30,23 @@ static unique_ptr<FunctionData> ReadCSVBind(ClientContext &context, TableFunctio
...
-		} else if (loption == "timestampformat") {
-			options.has_format[LogicalTypeId::TIMESTAMP] = true;
-			auto &timestamp_format = options.date_format[LogicalTypeId::TIMESTAMP];
-			timestamp_format.format_specifier = StringValue::Get(kv.second);
-			string error = StrTimeFormat::ParseFormatSpecifier(timestamp_format.format_specifier, timestamp_format);
-			if (!error.empty()) {
-				throw InvalidInputException("Could not parse TIMESTAMPFORMAT: %s", error.c_str());
+		set.clear();
+		set.emplace_back(kv.second);
+		if (options.SetReadOption(loption, set, names)) {
```
Comment (Mytherin) [inline src/function/table/read_csv.cpp:38]: Perhaps make a wrapper function that takes a single value as parameter (`options.SetReadOption(..., Value val)`)

File: src/function/table/read_csv.cpp [hunk 2]
```diff
@@ -30,78 +30,23 @@ static unique_ptr<FunctionData> ReadCSVBind(ClientContext &context, TableFunctio
...
+			continue;
+		}
+		if (loption == "columns") {
```
Comment (Mytherin) [inline src/function/table/read_csv.cpp:41]: If we check these special cases before calling `options.SetReadOption` we can immediately throw an exception in the `SetReadOption` function when something goes wrong.

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some feedback:

File: src/execution/operator/persistent/buffered_csv_reader.cpp [hunk 3]
```diff
@@ -149,6 +214,125 @@ void BufferedCSVReaderOptions::SetDelimiter(const string &input) {
...
+		escape = StringValue::Get(set[0]);
+		has_escape = true;
+	} else if (loption == "all_varchar") {
+		all_varchar = ParseBoolean(set);
+	} else if (loption == "normalize_names") {
+		normalize_names = ParseBoolean(set);
+	} else if (loption == "filename") {
+		include_file_name = ParseBoolean(set);
+	} else {
+		return false;
```
[the remark this answers is not in the document]
Comment (Tishj) [reply src/execution/operator/persistent/buffered_csv_reader.cpp:283]: Yes I'll do that, we dont have the original option here though, only the lowercase-converted one
Which might cause slight confusion

File: src/execution/operator/persistent/buffered_csv_reader.cpp [hunk 4]
```diff
@@ -149,6 +214,125 @@ void BufferedCSVReaderOptions::SetDelimiter(const string &input) {
...
+		has_header = true;
+	} else if (loption == "null" || loption == "nullstr") {
+		null_str = ParseString(set);
+	} else if (loption == "encoding") {
+		auto encoding = StringUtil::Lower(ParseString(set));
+		if (encoding != "utf8" && encoding != "utf-8") {
+			throw BinderException("Copy is only supported for UTF-8 encoded files, ENCODING 'UTF-8'");
+		}
+	} else if (loption == "compression") {
+		compression = FileCompressionTypeFromString(ParseString(set));
```
[the remark this answers is not in the document]
Comment (Tishj) [reply src/execution/operator/persistent/buffered_csv_reader.cpp:324]: Ah alright, I encountered them in ParseBaseOption so I left them as Common
I'll verify that they are read-only and move them to their appropriate category

[the remark this answers is not in the document]
Comment (Tishj) [reply src/execution/operator/persistent/buffered_csv_reader.cpp:324]:
```
static unique_ptr<GlobalFunctionData> WriteCSVInitializeGlobal(ClientContext &context, FunctionData &bind_data,
                                                               const string &file_path) {
	auto &csv_data = (WriteCSVData &)bind_data;
	auto &options = csv_data.options;
	auto global_data = make_unique<GlobalWriteCSVData>(FileSystem::GetFileSystem(context), file_path,
	                                                   FileSystem::GetFileOpener(context), options.compression);
	```
	
	This seems to suggest that `compression` is also used in Write

[the remark this answers is not in the document]
Comment (Tishj) [reply src/execution/operator/persistent/buffered_csv_reader.cpp]: `maximum_line_size` and `skip_rows` have been moved to the Read options, `compression` is kept as Common for now

**Review by Mytherin — APPROVED** [review]
Thanks for the updates! Looks good. Could you solve the merge conflicts?

**Comment by Tishj** [comment]
Merge conflicts resolved

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**